### DISCUSSION TITLE
what mean "Since it is the consumers instead of the broker who dispatch the messages from the batches"?

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#19516] [General]
https://pulsar.apache.org/docs/next/concepts-messaging/#batching-for-key-shared-subscriptions

`Since it is the consumers instead of the broker who dispatch the messages from the batches, the key of the first message in one batch is considered as the key to all messages in this batch, thereby leading to context errors`

mean "consumers dispatch message to other consumer"??

**Comment by startjava** [comment]
@momo-jun 
https://forum.streamnative.io/t/topic/592
two same question，N day, no answer .thank you

**Reply by momo-jun** [reply]
Sorry for the late response. From the context (for batched messages), your understanding seems correct. I'd like to ping engineers to confirm this point. 
@shibd Can you please help confirm the logic here?

**Reply by shibd** [reply]
> mean "consumers dispatch message to other consumer"??

No, The `consumers` here refer to the [objects](https://github.com/shibd/pulsar/blob/c875977340eb76cab62de68442ba6238866d9643/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L57) used inside the broker to dispatch messages.

>  the key of the first message in one batch is considered as the key to all messages in this batch, thereby leading to context errors

You can read this passage. In `batchMsg`, the broker will only extract the key of the first message to [select the consumer](https://github.com/apache/pulsar/blob/c875977340eb76cab62de68442ba6238866d9643/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L218-L219). Therefore, when the message's key in the batch is inconsistent, the consumer may receive the inconsistent key message.
